File: python/cnpj.py
from math import ceil


class CNPJ:
    def is_valid(self, in_cnpj: str) -> bool:
        if not in_cnpj:
            return False

        cnpj_invalidos = ["00000000000000"]

        try:
            self.cnpj = self.__remove_pontuacao(in_cnpj)

            if self.cnpj in cnpj_invalidos:
                return False

            # Remove os dígitos verificadores do CNPJ
            cnpj_sem_dv = self.__remove_dv_cnpj()
            # Calcula os dígitos verficadores
            dv = self.__gera_dv()
            return "%s%s" % (cnpj_sem_dv, dv) == self.cnpj
        except ValueError:
            return False


    def __gera_dv(self) -> str:
        cnpj_sem_dv = self.__remove_dv_cnpj()
        dv = DigitoVerificadorCNPJ(cnpj_sem_dv)
        return dv.calcula_dv()

    def __remove_dv_cnpj(self) -> str:
        if len(self.cnpj) == 14:
            return self.cnpj[0:-2]
        elif len(self.cnpj) == 12:
            return self.cnpj
        else:
            raise ValueError("CNPJ com tamanho inválido!")

    @staticmethod
    def __remove_pontuacao(in_cnpj: str) -> str:
        return ''.join(x for x in in_cnpj if x not in "./-")


class DigitoVerificadorCNPJ:
    def __init__(self, input: str):
        self._cnpj = input.upper()
        self._cnpj_dv = input.upper()

    def calcula_dv(self) -> str:
        dv1 = self.__calcula_dv1()
        dv2 = self.__calcula_dv2()
        return f"{dv1}{dv2}"

    def __calcula_dv1(self) -> int:
        dv = self.__calcula_soma_dv()
        self._cnpj_dv = f"{self._cnpj_dv}{dv}"
        return dv

    def __calcula_dv2(self) -> int:
        dv = self.__calcula_soma_dv()
        self._cnpj_dv = f"{self._cnpj_dv}{dv}"
        return dv

    def calcula_ascii(self, caracter: str):
        return ord(caracter) - 48

    def __calcula_soma_dv(self) -> int:
        tamanho_range = len(self._cnpj_dv)
        num_range = ceil(tamanho_range / 8)

        pesos = list()
        for i in range(num_range):
            pesos.extend(range(2, 10))

        pesos = pesos[0:tamanho_range]
        pesos.reverse()
        soma_produto = sum(a * b for a, b in zip(map(self.calcula_ascii, self._cnpj_dv), pesos))

        mod_soma = soma_produto % 11
        if mod_soma < 2:
            return 0
        else:
            return 11 - mod_soma

File: typescript/cnpj.ts

class CNPJ {
    private cnpj: string;

    private cnpjDV: string;

    isValid(inCNPJ: string): boolean {
        if (!inCNPJ) {
            return false;
        }

        const cnpjInvalidos = ["00000000000000"];

        try {
            this.cnpj = CNPJ.removePontuacao(inCNPJ);

            if (cnpjInvalidos.includes(this.cnpj)) {
                return false;
            }

            // Remove os dígitos verificadores do CNPJ
            const cnpjSemDV = this.removeDVCNPJ();
            // Calcula os dígitos verificadores
            const dv = this.geraDV();
            return `${cnpjSemDV}${dv}` === this.cnpj;
        } catch (error) {
            return false;
        }
    }

    private geraDV(): string {
        this.cnpjDV = this.removeDVCNPJ();
        return this.calculaDV();
    }

    private removeDVCNPJ(): string {
        if (this.cnpj.length === 14) {
            return this.cnpj.slice(0, -2);
        } else if (this.cnpj.length === 12) {
            return this.cnpj;
        } else {
            throw new Error("CNPJ com tamanho inválido!");
        }
    }

    calculaDV(): string {
        const dv1 = this.calculaDV1();
        const dv2 = this.calculaDV2();
        return `${dv1}${dv2}`;
    }

    private calculaDV1(): number {
        const dv = this.calculaSomaDV();
        this.cnpjDV = `${this.cnpjDV}${dv}`;
        return dv;
    }

    private calculaDV2(): number {
        const dv = this.calculaSomaDV();
        this.cnpjDV = `${this.cnpjDV}${dv}`;
        return dv;
    }

    private calculaAscii(caracter: string): number {
        return caracter.charCodeAt(0) - 48;
    }

    private calculaSomaDV(): number {
        const tamanhoRange = this.cnpjDV.length;
        const numRange = Math.ceil(tamanhoRange / 8);

        let pesos: number[] = [];
        for (let i = 0; i < numRange; i++) {
            pesos = pesos.concat([2, 3, 4, 5, 6, 7, 8, 9]);
        }

        pesos = pesos.slice(0, tamanhoRange).reverse();
        const somaProduto = Array.from(this.cnpjDV).reduce(
            (acc, char, index) => acc + this.calculaAscii(char) * pesos[index],
            0
        );

        const modSoma = somaProduto % 11;
        return modSoma < 2 ? 0 : 11 - modSoma;
    }

    private static removePontuacao(inCNPJ: string): string {
        return inCNPJ.replace(/[.\-/]/g, '');
    }
}

